File: 人工智能概论/exp1/items.py
import pygame
import random
from constants import *
from utils import load_image

class Item:
    def __init__(self, item_type, position):
        self.type = item_type
        self.position = position
        self.value = ITEM_VALUES[item_type]
        self.weight = ITEM_WEIGHTS[item_type]
        self.size = ITEM_SIZES[item_type]
        self.caught = False
        self.rect = pygame.Rect(0, 0, self.size, self.size)
        self.rect.center = position
        
        # 加载图片或创建默认图形
        self.image = self._load_image()
    
    def _load_image(self):
        """根据物品类型加载对应的图片"""
        image_paths = {
            ITEM_GOLD_SMALL: "resources/images/gold/small.png",
            ITEM_GOLD_MEDIUM: "resources/images/gold/medium.png",
            ITEM_GOLD_LARGE: "resources/images/gold/large.png",
            ITEM_STONE_SMALL: "resources/images/stone/small.png",
            ITEM_STONE_MEDIUM: "resources/images/stone/medium.png",
            ITEM_STONE_LARGE: "resources/images/stone/large.png",
            ITEM_DIAMOND: "resources/images/diamond.png",
            ITEM_BOMB: "resources/images/bomb.png"
        }
        
        colors = {
            ITEM_GOLD_SMALL: GOLD,
            ITEM_GOLD_MEDIUM: GOLD,
            ITEM_GOLD_LARGE: GOLD,
            ITEM_STONE_SMALL: GRAY,
            ITEM_STONE_MEDIUM: GRAY,
            ITEM_STONE_LARGE: GRAY,
            ITEM_DIAMOND: BLUE,
            ITEM_BOMB: RED
        }
        
        path = image_paths.get(self.type, "")
        color = colors.get(self.type, WHITE)
        size = (self.size, self.size)
        
        return load_image(path, color, size)
    
    def draw(self, screen):
        """绘制物品"""
        screen.blit(self.image, self.rect)
        
        # 如果被钩住，显示价值
        if self.caught:
            font = pygame.font.SysFont(None, 24)
            text = font.render(f"${self.value}", True, WHITE)
            screen.blit(text, (self.rect.centerx - text.get_width() // 2, 
                              self.rect.top - 20))

class ItemManager:
    def __init__(self):
        self.items = []
        self.caught_items = []
    
    def generate_level_items(self, level):
        """根据关卡生成物品"""
        self.items = []
        
        # 根据关卡难度调整物品数量和类型
        num_items = 10 + level * 2
        
        for _ in range(num_items):
            item_type = self._get_random_item_type(level)
            position = self._get_valid_position()
            self.items.append(Item(item_type, position))
    
    def _get_random_item_type(self, level):
        """根据关卡随机生成物品类型"""
        # 随着关卡提高，高价值物品出现概率增加
        probabilities = {
            ITEM_GOLD_SMALL: 0.3 - level * 0.02,
            ITEM_GOLD_MEDIUM: 0.2 + level * 0.01,
            ITEM_GOLD_LARGE: 0.1 + level * 0.01,
            ITEM_STONE_SMALL: 0.2 - level * 0.02,
            ITEM_STONE_MEDIUM: 0.1 - level * 0.01,
            ITEM_STONE_LARGE: 0.05,
            ITEM_DIAMOND: 0.02 + level * 0.005,
            ITEM_BOMB: 0.03
        }
        
        # 确保概率总和为1
        total = sum(probabilities.values())
        for key in probabilities:
            probabilities[key] /= total
        
        rand = random.random()
        cumulative = 0
        
        for item_type, prob in probabilities.items():
            cumulative += prob
            if rand <= cumulative:
                return item_type
        
        return ITEM_GOLD_SMALL
    
    def _get_valid_position(self):
        """获取有效的物品位置（不重叠）"""
        max_attempts = 100
        
        for _ in range(max_attempts):
            x = random.randint(100, SCREEN_WIDTH - 100)
            y = random.randint(200, SCREEN_HEIGHT - 100)
            
            # 检查是否与其他物品重叠
            valid = True
            new_rect = pygame.Rect(0, 0, 50, 50)
            new_rect.center = (x, y)
            
            for item in self.items:
                if new_rect.colliderect(item.rect):
                    valid = False
                    break
            
            if valid:
                return (x, y)
        
        # 如果找不到有效位置，返回随机位置
        return (random.randint(100, SCREEN_WIDTH - 100), 
                random.randint(200, SCREEN_HEIGHT - 100))
    
    def check_collision(self, hook_rect):
        """检查钩子与物品的碰撞"""
        for item in self.items:
            if not item.caught and hook_rect.colliderect(item.rect):
                item.caught = True
                self.caught_items.append(item)
                return item
        
        return None
    
    def draw(self, screen):
        """绘制所有物品"""
        for item in self.items:
            item.draw(screen)
    
    def remove_caught_items(self):
        """移除已被捕获的物品"""
        self.items = [item for item in self.items if not item.caught]
        self.caught_items = []

File: 人工智能概论/exp1/gold-miner-v2/js/items.js
// 物品系统
class Item {
    constructor(type, position) {
        this.type = type;
        this.position = position;
        this.value = ITEM_PROPERTIES[type].value;
        this.weight = ITEM_PROPERTIES[type].weight;
        this.size = ITEM_PROPERTIES[type].size;
        this.color = ITEM_PROPERTIES[type].color;
        this.caught = false;
        this.image = null;
    }
    
    async loadImage() {
        const imagePaths = {
            [ITEM_TYPES.GOLD_SMALL]: 'assets/images/gold-small.png',
            [ITEM_TYPES.GOLD_MEDIUM]: 'assets/images/gold-medium.png',
            [ITEM_TYPES.GOLD_LARGE]: 'assets/images/gold-large.png',
            [ITEM_TYPES.STONE_SMALL]: 'assets/images/stone-small.png',
            [ITEM_TYPES.STONE_MEDIUM]: 'assets/images/stone-medium.png',
            [ITEM_TYPES.STONE_LARGE]: 'assets/images/stone-large.png',
            [ITEM_TYPES.DIAMOND]: 'assets/images/diamond.png',
            [ITEM_TYPES.BOMB]: 'assets/images/bomb.png'
        };
        
        this.image = await Utils.loadImage(
            imagePaths[this.type], 
            this.color,
            {width: this.size, height: this.size}
        );
    }
    
    draw(ctx) {
        if (this.image) {
            // 使用图片绘制物品
            ctx.drawImage(
                this.image,
                this.position.x - this.size / 2,
                this.position.y - this.size / 2,
                this.size,
                this.size
            );
        } else {
            // 使用颜色绘制物品
            ctx.fillStyle = this.color;
            ctx.beginPath();
            ctx.arc(this.position.x, this.position.y, this.size / 2, 0, Math.PI * 2);
            ctx.fill();
            ctx.strokeStyle = '#000';
            ctx.lineWidth = 2;
            ctx.stroke();
            
            // 特殊物品的标识
            if (this.type === ITEM_TYPES.BOMB) {
                // 炸弹绘制"X"标记
                ctx.strokeStyle = '#000';
                ctx.lineWidth = 3;
                ctx.beginPath();
                ctx.moveTo(this.position.x - this.size/3, this.position.y - this.size/3);
                ctx.lineTo(this.position.x + this.size/3, this.position.y + this.size/3);
                ctx.moveTo(this.position.x + this.size/3, this.position.y - this.size/3);
                ctx.lineTo(this.position.x - this.size/3, this.position.y + this.size/3);
                ctx.stroke();
            } else if (this.type === ITEM_TYPES.DIAMOND) {
                // 钻石绘制特殊形状
                ctx.fillStyle = '#FFFFFF';
                ctx.font = 'bold 12px Arial';
                ctx.textAlign = 'center';
                ctx.textBaseline = 'middle';
                ctx.fillText('💎', this.position.x, this.position.y);
            } else if (this.type.includes('stone')) {
                // 石头绘制纹理
                ctx.fillStyle = '#606060';
                ctx.beginPath();
                ctx.arc(this.position.x - 3, this.position.y - 3, 3, 0, Math.PI * 2);
                ctx.arc(this.position.x + 4, this.position.y + 2, 4, 0, Math.PI * 2);
                ctx.arc(this.position.x - 2, this.position.y + 4, 3, 0, Math.PI * 2);
                ctx.fill();
            } else if (this.type.includes('gold')) {
                // 黄金绘制光泽
                ctx.fillStyle = 'rgba(255, 255, 255, 0.4)';
                ctx.beginPath();
                ctx.arc(this.position.x - this.size/4, this.position.y - this.size/4, this.size/5, 0, Math.PI * 2);
                ctx.fill();
            }
        }
        
        // 如果被钩住，显示价值
        if (this.caught) {
            ctx.fillStyle = '#FFF';
            ctx.font = '16px Arial';
            ctx.textAlign = 'center';
            ctx.textBaseline = 'middle';
            ctx.fillText(`$${this.value}`, this.position.x, this.position.y - this.size / 2 - 10);
        }
    }
    
    getRect() {
        return {
            x: this.position.x - this.size / 2,
            y: this.position.y - this.size / 2,
            width: this.size,
            height: this.size
        };
    }
}

// 物品管理器
class ItemManager {
    constructor() {
        this.items = [];
    }
    
    async generateLevelItems(level) {
        this.items = [];
        const levelConfig = LEVEL_CONFIG[level - 1];
        const numItems = levelConfig.itemCount;
        
        // 创建物品
        for (let i = 0; i < numItems; i++) {
            const itemType = this.getRandomItemType(level);
            const position = this.getValidPosition();
            const item = new Item(itemType, position);
            await item.loadImage();
            this.items.push(item);
        }
    }
    
    getRandomItemType(level) {
        const levelConfig = LEVEL_CONFIG[level - 1];
        const probabilities = levelConfig.probabilities;
        
        // 确保概率总和为1
        const total = Object.values(probabilities).reduce((sum, prob) => sum + prob, 0);
        const normalizedProbabilities = {};
        Object.keys(probabilities).forEach(key => {
            normalizedProbabilities[key] = probabilities[key] / total;
        });
        
        const rand = Math.random();
        let cumulative = 0;
        
        for (const itemType in normalizedProbabilities) {
            cumulative += normalizedProbabilities[itemType];
            if (rand <= cumulative) {
                return itemType;
            }
        }
        
        return ITEM_TYPES.GOLD_SMALL;
    }
    
    getValidPosition() {
        const maxAttempts = 100;
        
        for (let i = 0; i < maxAttempts; i++) {
            const x = Utils.randomInt(100, CANVAS_WIDTH - 100);
            const y = Utils.randomInt(200, CANVAS_HEIGHT - 100);
            
            // 检查是否与其他物品重叠
            let valid = true;
            const newRect = {
                x: x - 25,
                y: y - 25,
                width: 50,
                height: 50
            };
            
            for (const item of this.items) {
                const itemRect = item.getRect();
                if (this.checkRectCollision(newRect, itemRect)) {
                    valid = false;
                    break;
                }
            }
            
            if (valid) {
                return { x, y };
            }
        }
        
        // 如果找不到有效位置，返回随机位置
        return {
            x: Utils.randomInt(100, CANVAS_WIDTH - 100),
            y: Utils.randomInt(200, CANVAS_HEIGHT - 100)
        };
    }
    
    checkRectCollision(rect1, rect2) {
        return rect1.x < rect2.x + rect2.width &&
               rect1.x + rect1.width > rect2.x &&
               rect1.y < rect2.y + rect2.height &&
               rect1.y + rect1.height > rect2.y;
    }
    
    checkHookCollision(hookRect) {
        for (const item of this.items) {
            if (!item.caught && this.checkRectCollision(hookRect, item.getRect())) {
                item.caught = true;
                return item;
            }
        }
        return null;
    }
    
    removeItem(item) {
        this.items = this.items.filter(i => i !== item);
    }
    
    draw(ctx) {
        for (const item of this.items) {
            item.draw(ctx);
        }
    }
}